internal/ledgerutil/compare.go

```diff
@@ -283,51 +282,63 @@ func findAndWriteDifferences(outputDirPath string, outputFilename string, snapsh
 }
 
 // firstRecords is a struct used to hold only the earliest records up to a given limit
+// Basically a max heap with a size limit
 type firstRecords struct {
-	records       *diffRecordSlice
-	highestRecord *diffRecord
-	highestIndex  int
-	limit         int
+	records *diffRecordHeap
+	limit   int
 }
 
 func (s *firstRecords) addRecord(r *diffRecord) {
+	if s.limit == 0 {
+		return
+	}
+	// Limit not reached, can still add elements
 	if len(*s.records) < s.limit {
-		*s.records = append(*s.records, r)
-		s.setHighestRecord()
+		heap.Push(s.records, r)
+		// Limit reached, check new record against root
 	} else {
-		if r.lessThan(s.highestRecord) {
-			(*s.records)[s.highestIndex] = r
-			s.setHighestRecord()
+		// Only add records earlier than the root
+		if r.lessThan((*s.records)[0]) {
+			heap.Push(s.records, r)
+			heap.Pop(s.records)
 		}
 	}
 }
 
-func (s *firstRecords) setHighestRecord() {
-	if len(*s.records) == 1 {
-		s.highestRecord = (*s.records)[0]
-		s.highestIndex = 0
-		return
-	}
-	for i, r := range *s.records {
-		if s.highestRecord.lessThan(r) {
-			s.highestRecord = r
-			s.highestIndex = i
-		}
+// Returns list of diffRecords currently in heap sorted from earliest to latest, will empty the heap
+func (s *firstRecords) getAllRecords() []*diffRecord {
+	n := len(*s.records)
+	var t []*diffRecord
+	for i := 0; i < n; i++ {
+		r := heap.Pop(s.records)
+		t = append([]*diffRecord{r.(*diffRecord)}, t...)
```

Comment:
As you know the size in advance, you can avoid repetitive memory allocation. 
```suggestion
		t := make([]*diffRecord, n)
	for i := 0; i < n; i++ {
		r := heap.Pop(s.records)
		t[n-i-1] = r.(*diffRecord)
	}
```

Reply:
Makes sense 👍🏼